File: 22.generate-parentheses/solution.cpp
class Solution {
public:
	vector<string> generateParenthesis(int n) {
		vector<string> result;
		generateParenthesis(result, "", n, n);
		return result;
	}

private:
	void generateParenthesis(vector<string> &result, string temp, int l, int r)
	{
		if (r == 0)
		{
			result.push_back(temp);
			return;
		}
		if (l == 0)
		{
			string s;
			for (int i = 0; i < r; i++)
			{
				s += ')';
			}
			result.push_back(temp + s);
			return;
		}
		generateParenthesis(result, temp + "(", l - 1, r);
		if (l < r)
		{
			generateParenthesis(result, temp + ")", l, r - 1);
		}
	}
};

File: 22.generate-parentheses/solution.py
class Solution:
    def __init__(self):
        self.result = []

    def generateParenthesis(self, n: int) -> List[str]:
        self.result = []
        self.generateParenthesis2(n, 0, 0, "")
        return self.result

    def generateParenthesis2(self, n: int, left: int, right: int, cur: str):
        if left == n and right == n:
            self.result.append(cur)
        if left < n:
            self.generateParenthesis2(n, left+1, right, cur+"(")
        if left > right:
            self.generateParenthesis2(n, left, right+1, cur+")")
